plugins/woocommerce/tests/php/includes/class-wc-order-factory-test.php

```diff
@@ -98,4 +101,42 @@ public function test_cache_dont_interfere_with_orders() {
 		OrderHelper::toggle_cot_feature_and_usage( false );
 	}
 
+	/**
+	 * @testDox Test that an WC_Order_Factory::get_order() does not return an order that did not properly load, $id = 0.
+	 */
+	public function test_get_order_doesnt_return_invalid_cached_order() {
+		global $wpdb;
+
+		OrderHelper::toggle_cot_feature_and_usage( true );
+		$this->setup_serializing_cache();
+
+		$order = OrderHelper::create_order();
+
+		// Retrieve the order to prime the cache.
+		$retrieved_order = WC_Order_Factory::get_order( $order->get_id() );
+		$this->assertEquals( $order->get_id(), $retrieved_order->get_id() );
+
+		// Delete the order from the DB to mimic situations like SQL replication lag where the cache has propagated,
+		// but SQL data hasn't yet caught up.
+		$wpdb->delete( OrdersTableDataStore::get_orders_table_name(), array( 'ID' => $retrieved_order->get_id() ) );
+		foreach ( OrdersTableDataStore::get_all_table_names_with_id() as $table_id => $table ) {
+			if ( 'orders' !== $table_id ) {
+				$wpdb->delete( $table, array( 'order_id' => $retrieved_order->get_id() ) );
+			}
+		}
+
+		/**
+		 * Retrieving the order again should either:
+		 * - return false because the order was not able to fully load.
+		 * - return a fully loaded order. Our direct DB write purposefully does not trigger a cache refresh so a fully
+		 *      loaded order should also be an expected result with cache enabled.
+		 */
+		$retrieved_order = WC_Order_Factory::get_order( $order->get_id() );
+		if ( is_object( $retrieved_order ) ) {
+			$this->assertEquals( $order->get_id(), $retrieved_order->get_id() );
+		} else {
+			$this->assertFalse( $retrieved_order );
+		}
```

Comment:
It looks like this should be two different tests, one in which the order exists in the db and is then retrieved successfully, and another one in which the order only exists in the cache and `false` is returned. Maybe you can use [PHPUnit's testWith annotation](https://docs.phpunit.de/en/9.6/annotations.html#testwith) to pass an `$orderExistsInDb` boolean to the test method and thus reuse the same method for both test cases.

Reply:
@Konamiman The test here is verifying that `WC_Order_Factory::get_order()` doesn't return an Order object with and ID of `0`.  The behavior here, though, depends on whether #46023 is merged. 

Without #46023, ``WC_Order_Factory::get_order()` should return false because the order data wasn't able to be loaded from the database.  However, with #46023, the order would still be able to load because the data is available in cache.  So, if one of this or #46023 is merged, this test will break for the other without the if/else here.

I'll just assume this simpler PR would be merged first, and check that the `$retrieved_order` is always `false` and then update #46023 with two tests based on cache state.

Reply:
> I'll just assume this simpler PR would be merged first, and check that the $retrieved_order is always false and then update https://github.com/woocommerce/woocommerce/pull/46023 with two tests based on cache state.

Thanks, that makes sense; as long as you remember to rebase https://github.com/woocommerce/woocommerce/pull/46023 from `trunk` after this one is merged, we're good.

Good job with `SerializingCacheProxy` and `SerializingCacheTrait`, by the way; any helpful additions and improvements on the unit testing infrastructure are highly welcome.